fms_extras/utils/generation.py

```diff
@@ -3,12 +3,344 @@
+            the indices used in performing flattening
+
+    Returns:
+    PagedAttentionCacheData
+        the cache data with proper unflattening metadata included
+    """
+    cache_data.unflatten_indices = unflat_indices
+    cache_data.flatten_indices = flat_indices
+    cache_data.position_ids = apply_index_map(
+        cache_data.position_ids.view(-1), flat_indices
+    )[
+        None,
+    ]
+    cache_data.context_lengths = apply_index_map(
+        cache_data.context_lengths, cache_data.flatten_indices  # type: ignore
+    )  # n'
+    cache_data.block_mapping = apply_index_map(
+        cache_data.block_mapping, cache_data.flatten_indices
+    )  # n' n_blocks
+
+    return cache_data
+
+
+def __perform_correctness_checks(
+    input_ids_unflat: torch.Tensor,
+    next_vals: torch.Tensor,
+    embeds: torch.Tensor,
+    batch_size: int,
+    num_candidates_per_sequence: int,
+    decode_seq_length: int,
+) -> Tuple[torch.Tensor, torch.Tensor, torch.Tensor, torch.Tensor]:
+    """
+    Check for the correctness of speculator predictions and get the best candidates as well as the number of tokens
+    correct for each of those candidates
+
+    Args:
+        input_ids_unflat: torch.Tensor
+            the original input ids unflattened
+        next_vals: torch.Tensor
+            a tensor of the next tokens
+        embeds: torch.Tensor
+            the output embeddings for the best guesses
+        batch_size: int
+            the number of sequences in the batch
+        num_candidates_per_sequence: int
+            the number of candidates created per sequence
+        decode_seq_length: int
+            the model input length at decode step
+
+    Returns:
+    Tuple[torch.Tensor, torch.Tensor, torch.Tensor, torch.Tensor]
+        a tensor of the next tokens per best guess, a tensor of the embeds per best guess, a tensor of the number of
+        correct tokens per best guess, and a tensor containing the best guess amongst all candidates per sequence
+    """
+    # Check correctness of speculator predictions
+    test = input_ids_unflat[:, :, 1:].eq(next_vals[:, :, :-1]).cumprod(2)
+    n_correct = test.sum(2).view(batch_size, num_candidates_per_sequence)
+    best_guess = n_correct.argmax(1)  # b
+    best_guess_unflat = (
+        best_guess.unsqueeze(1).expand(batch_size, decode_seq_length).unsqueeze(1)
+    )  # b 1 1+h
+
+    # Set global values to those of best guess
+    next_vals = next_vals.gather(1, best_guess_unflat).squeeze(1)  # b 1+h
+    n_correct = n_correct.gather(1, best_guess.unsqueeze(1)).squeeze(1)  # b
+    embeds = embeds.gather(
+        1, best_guess_unflat.unsqueeze(3).expand(-1, -1, -1, embeds.size(3))
+    ).squeeze(
+        1
+    )  # b 1+h d
+    return next_vals, embeds, n_correct, best_guess
+
+
+def __get_correct_tokens(
+    next_vals: torch.Tensor, n_correct: torch.Tensor, embeds: torch.Tensor
+) -> Tuple[List[torch.Tensor], torch.Tensor]:
+    """
+    get the correct tokens and embeds
```

Comment:
Extract the last correct token/embedding from a single candidate, to be used as the start of the next set of speculative candidates

Reply:
fixed in latest commit